dlc/src/digit_decomposition.rs

```diff
@@ -0,0 +1,673 @@
+//! Utility functions to decompose numeric outcome values
+
+use super::{Error, Payout, RangeOutcome};
+
+struct PrefixRange {
+    start: Vec<usize>,
+    end: Vec<usize>,
+    prefix: Vec<usize>,
+}
+
+/// Decompose a numeric value into digits in the specified base. If the decomposed
+/// value contains less than `nb_digits`, zeroes will be prepended to reach `nb_digits`
+/// size.
+pub fn decompose_value(mut value: usize, base: usize, nb_digits: usize) -> Vec<usize> {
+    let mut res = Vec::new();
+
+    while value > 0 {
+        res.push(value % base);
+        value = ((value as f64) / (base as f64)).floor() as usize;
+    }
+
+    while res.len() < nb_digits {
+        res.push(0);
+    }
+
+    res.into_iter().rev().collect()
+}
+
+/// Takes a decomposed representation of a numerical value in a given base and returns
+/// the represented value as a `usize`
+pub fn compose_value(values: &[usize], base: usize) -> usize {
+    let mut composed = 0;
+    for i in 0..values.len() {
+        let pow = values.len() - i - 1;
+        composed += values[i] * base.pow(pow as u32);
+    }
+
+    composed
+}
+
+/// Takes a vector or `RangeOutcome` and (if necessary) adds one at the beginning
+/// and at the end of it to cover the full range of possible outcome values
+pub fn get_max_ranges(
```

Comment:
perhaps a more descriptive name would be `pad_...` this name seems really misleading, it took me a while to figure out what this function was doing. (also the doc description of "adds one at the beginning and at the end" didn't mean much because I had to read all the code to figure out the answer to "one of what" and also it isn't adding, it is modifying the extremal ranges to extend to the boundaries).

Reply:
Good point. Went with `pad_range_payouts` and tried to write a better doc for it. Also the doc was definitely wrong about the adding part thanks for spotting!